lib/pcb/svg-object-fns/create-svg-objects-from-pcb-rat-nests.ts

```diff
@@ -0,0 +1,149 @@
+import { getFullConnectivityMapFromCircuitJson } from "circuit-json-to-connectivity-map"
+import type { AnyCircuitElement } from "circuit-json"
+import { type INode as SvgObject } from "svgson"
+import { type Matrix, applyToPoint } from "transformation-matrix"
+
+interface RatsNestLine {
+  key: string
+  startPoint: { x: number; y: number }
+  endPoint: { x: number; y: number }
+  isInNet: boolean
+}
+
+export function createSvgObjectsForRatsNest(
+  soup: AnyCircuitElement[],
+  transform: Matrix,
+): SvgObject[] {
+  // Build a lookup map by cleaned ID for any element that might provide a position.
+  const elementMap = new Map<string, AnyCircuitElement>()
+  for (const elm of soup) {
+    const id =
+      (elm as any).id || (elm as any).pcb_port_id || (elm as any).source_port_id
+    if (id) {
+      const cleanedId = id.replace(/_\d+$/, "")
+      elementMap.set(cleanedId, elm)
+    }
+  }
+
+  const getElementPosition = (id: string): { x: number; y: number } | null => {
+    const cleanedId = id.replace(/_\d+$/, "")
+    const elm = elementMap.get(cleanedId)
+    if (elm && "x" in elm && "y" in elm) {
+      return { x: (elm as any).x, y: (elm as any).y }
+    }
+    return null
+  }
+
+  // Compute connectivity using the helper from the imported package.
+  let netMap: { [netId: string]: string[] } = {}
+  let idToNetMap: { [id: string]: string } = {}
+  try {
+    const connectivity = getFullConnectivityMapFromCircuitJson(soup)
+    netMap = connectivity.netMap
+    idToNetMap = connectivity.idToNetMap
+  } catch (error) {
+    console.error("Error computing connectivity map for rats nest:", error)
+    return []
+  }
+
+  // Filter for ports and traces that are relevant for rats nest.
+  const pcbPorts = soup.filter((elm) => elm.type === "pcb_port")
+  const sourceTraces = soup.filter((elm) => elm.type === "source_trace")
+  const sourcePorts = soup.filter((elm) => elm.type === "source_port")
+
+  const getSourcePortForPcbPort = (pcbPortId: string) => {
+    return sourcePorts.find((elm) => (elm as any).pcb_port_id === pcbPortId)
```

Comment:
Can you use `@tscircuit/soup-util` for this? Yes the name is weird, but it has much much better types and also checks for element types properly and auto-optimizes

```suggestion
    return su(circuitJson).source_port.getWhere({ pcb_port_id })
```

Generally you want to use soup-util instead of scanning the circuit json because it'll auto-optimize and not get slow even as the array gets massive. It has a similar interface to a db client

https://github.com/tscircuit/soup-util